imu/icm20xxx.c

```diff
@@ -0,0 +1,216 @@
+#define VAL_USER_CTRL_I2C_MST_RST 0x02
+
+/* ACCEL CONFIG */
+#define REG_ACCEL_CONFIG   0x14
+#define VAL_ACCEL_DLPCFG_6 0x30
+#define VAL_ACCEL_DLPCFG_4 0x20
+#define VAL_ACCEL_DLPCFG_2 0x10
+#define VAL_ACCEL_FS_2G    0x00
+#define VAL_ACCEL_FS_4G    0x02
+#define VAL_ACCEL_FS_8G    0x04
+#define VAL_ACCEL_FS_16G   0x06
+#define VAL_ACCEL_DLPF     0x01
+/* LSB for ACCEL sample rate div */
+#define REG_ACCEL_SMPLRT_DIV_2 0x11
+#define VAL_ACCEL_SMPLRT_DIV_2 0x07
+
+/* GYRO CONFIG */
+#define REG_GYR_CONFIG_1    0x01
+#define VAL_GYRO_DLPCFG_6   0x30
+#define VAL_GYRO_DLPCFG_4   0x20
+#define VAL_GYRO_DLPCFG_2   0x10
+#define VAL_GYRO_FS_DPS250  0x00
+#define VAL_GYRO_FS_DPS500  0x02
+#define VAL_GYRO_FS_DPS1000 0x04
+#define VAL_GYRO_FS_DPS2000 0x06
+#define VAL_GYRO_DLPF       0x01
+/* LSB for GYRO sample rate div */
+#define REG_GYR_SMPLRT_DIV_2 0x00
+#define VAL_GYR_SMPLRT_DIV_2 0x07
+
+/* data addresses */
+#define REG_DATA_ALL 0x2D /* actually ACEL_XOUT_H address */
+
+
+static immu_t imu_common = {
+	.pwr_mode = mode_sleep,
+	.scale_acc = g8,
+	.scale_gyr = dps1000
+};
+
+
+static int icm20xxx_reset(const i2c_dev_t *dev)
+{
+	int ret = 0;
+
+	ret += sendbyte(dev->addr, REG_BANK, VAL_BANK_0);           /* switch to user bank 0 */
+	ret += sendbyte(dev->addr, REG_PWR_MIGMT_1, VAL_PWR_RESET); /* reset device */
+	usleep(1000 * 10);
+
+	return (ret < 0) ? -EIO : EOK;
+}
+
+
+static int icm20xxx_run_mode(const i2c_dev_t *dev)
+{
+	int ret = 0;
+
+	ret += sendbyte(dev->addr, REG_BANK, VAL_BANK_0);         /* switch to user bank 0 */
+	ret += sendbyte(dev->addr, REG_PWR_MIGMT_1, VAL_PWR_RUN); /* put it in run mode */
+	usleep(1000 * 10);
+
+	/* set bypass mode for auxiliary I2C bus */
+	ret += sendbyte(dev->addr, REG_USER_CTRL, 0);
+	ret += sendbyte(dev->addr, 0x0f, 0 | (1 << 1));
+
+	if (ret >= 0) {
+		imu_common.pwr_mode = mode_full;
+		return EOK;
+	}
+	return -EIO;
+}
+
+
+static float translateAccel(short val)
+{
+	float fval = val / 32767.F;
+	switch (imu_common.scale_acc) {
+		case g2:
+			return fval * 2.F;
+		case g4:
+			return fval * 4.F;
+		case g16:
+			return fval * 16.F;
+		default:
+			return fval * 8.F;
+	}
+}
+
+
+static float translateGyr(short val)
+{
+	float fval = DEG2RAD * val / 32767;
+	switch (imu_common.scale_acc) {
+		case dps250:
+			return fval * 250;
+		case dps500:
+			return fval * 500;
+		case dps2000:
+			return fval * 2000;
+		default:
+			return fval * 1000;
+	}
+}
+
+
+static float translateTemp(short val)
+{
+	return 21.F + val / 333.87F;
+}
+
+
+int icm20xxx_init(const i2c_dev_t *dev)
+{
+	int ret = 0;
+
+	ret += icm20xxx_reset(dev);
+	ret += icm20xxx_run_mode(dev);
+	if (ret < 0) {
+		fprintf(stderr, "Cannot reset and put in run\n");
+		return -EIO;
+	}
+	ret = 0;
+
+	/* Accellerometer and Gyro configs are in USER_BANK_2 */
+	ret += sendbyte(dev->addr, REG_BANK, VAL_BANK_2);
+	{
+		/* set accel sample rate divider */
+		ret += sendbyte(dev->addr, REG_ACCEL_SMPLRT_DIV_2, VAL_ACCEL_SMPLRT_DIV_2);
```

Comment:
if the bank switching failed, we should probably not issue these commands (we might be changing other configs?)